ray-curator/ray_curator/stages/filters/heuristic_filter.py

```diff
@@ -0,0 +1,835 @@
+
+    def keep_document(self, score: float) -> bool:
+        return score <= self._cutoff
+
+
+class CommonEnglishWordsFilter(DocumentFilter):
+    """
+    If the sentence contains at least 2 common English words, then keep it.
+    NOTE: We purposefully check for the lowercase versions of those common words
+    to remove documents with over-capitalization.
+
+    For Chinese and Japanese text, we use external libraries to split the text
+    because these languages are not separated by spaces. For all other languages,
+    such as English, we assume words are separated by spaces.
+    """
+
+    def __init__(self, min_num_common_words: int = 2, stop_at_false: bool = True):
+        super().__init__()
+        self._cutoff = min_num_common_words
+        self._stop_at_false = stop_at_false
+        self._word_splitter = get_word_splitter("en")
+        self._name = "common_english_words"
+
+    def score_document(self, text: str) -> int:
+        common_word_counter = 0
+        for word in self._word_splitter(text.strip()):
+            if word in common_english_words:
+                common_word_counter += 1
+            if self._stop_at_false and common_word_counter >= self._cutoff:
+                return common_word_counter
+
+        return common_word_counter
+
+    def keep_document(self, score: int) -> bool:
+        return score >= self._cutoff
+
+
+class WordsWithoutAlphabetsFilter(DocumentFilter):
+    """
+    80% of words in a document must contain at least one alphabetic character.
+    Source: Gopher (Rae et al., 2021)
+
+    For Chinese and Japanese text, we use external libraries to split the text
+    because these languages are not separated by spaces. For all other languages,
+    such as English, we assume words are separated by spaces.
+    """
+
+    def __init__(self, min_words_with_alphabets: float = 0.8, lang: str = "en"):
+        super().__init__()
+        self._cutoff = min_words_with_alphabets
+        self._word_splitter = get_word_splitter(lang)
+        self._name = "words_without_alphabets"
+
+    def score_document(self, text: str) -> float:
+        num_english_alpha = 0
+        words = self._word_splitter(text.strip())
+        for word in words:
+            if regex_alpha.search(word):
+                num_english_alpha += 1
+
+        return num_english_alpha / len(words)
+
+    def keep_document(self, score: float) -> bool:
+        return score >= self._cutoff
+
+
+class PornographicUrlsFilter(DocumentFilter):
+    """
+    Check if any of the URLs within the document point to pornography.
+    """
+
+    def __init__(self):
+        super().__init__()
+
+    def score_document(self, text: str) -> int:
+        all_urls = regex_url.findall(text)
+        for url in all_urls:
+            if "porn" in url:
+                return 1
+
+        return 0
+
+    def keep_document(self, score: int) -> bool:
+        return score != 1
+
+
+class TokenCountFilter(DocumentFilter):
```

Comment:
We could improve performance for this function with `@batched`, since `AutoTokenizer` supports batch encoding. Ironically, Curator with Dask does not take advantage of this.

For now, I will track this in https://github.com/NVIDIA-NeMo/Curator/issues/706, which includes adding batch tokenization for custom filters in an existing Curator tutorial.